// @jsonforms/examples/lib/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
/*
  The MIT License

  Copyright (c) 2017-2019 EclipseSource Munich
  https://github.com/eclipsesource/jsonforms

  Permission is hereby granted, free of charge, to any person obtaining a copy
  of this software and associated documentation files (the "Software"), to deal
  in the Software without restriction, including without limitation the rights
  to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
  copies of the Software, and to permit persons to whom the Software is
  furnished to do so, subject to the following conditions:

  The above copyright notice and this permission notice shall be included in
  all copies or substantial portions of the Software.

  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
  THE SOFTWARE.
*/
var allOf = tslib_1.__importStar(require("./allOf"));
exports.allOf = allOf;
var anyOf = tslib_1.__importStar(require("./anyOf"));
exports.anyOf = anyOf;
var oneOf = tslib_1.__importStar(require("./oneOf"));
exports.oneOf = oneOf;
var oneOfArray = tslib_1.__importStar(require("./oneOfArray"));
exports.oneOfArray = oneOfArray;
var array = tslib_1.__importStar(require("./arrays"));
exports.array = array;
var nestedArray = tslib_1.__importStar(require("./nestedArrays"));
exports.nestedArray = nestedArray;
var arrayWithDetail = tslib_1.__importStar(require("./arrays-with-detail"));
exports.arrayWithDetail = arrayWithDetail;
var arrayWithDetailAndRule = tslib_1.__importStar(require("./arrays-with-detail-and-rule"));
exports.arrayWithDetailAndRule = arrayWithDetailAndRule;
var arrayWithCustomChildLabel = tslib_1.__importStar(require("./arrays-with-custom-element-label"));
exports.arrayWithCustomChildLabel = arrayWithCustomChildLabel;
var stringArray = tslib_1.__importStar(require("./stringArray"));
exports.stringArray = stringArray;
var categorization = tslib_1.__importStar(require("./categorization"));
exports.categorization = categorization;
var stepper = tslib_1.__importStar(require("./stepper"));
exports.stepper = stepper;
var steppershownav = tslib_1.__importStar(require("./steppershownav"));
exports.steppershownav = steppershownav;
var controlOptions = tslib_1.__importStar(require("./control-options"));
exports.controlOptions = controlOptions;
var day1 = tslib_1.__importStar(require("./day1"));
exports.day1 = day1;
var day2 = tslib_1.__importStar(require("./day2"));
exports.day2 = day2;
var day3 = tslib_1.__importStar(require("./day3"));
exports.day3 = day3;
var day4 = tslib_1.__importStar(require("./day4"));
exports.day4 = day4;
var day5 = tslib_1.__importStar(require("./day5"));
exports.day5 = day5;
var day6 = tslib_1.__importStar(require("./day6"));
exports.day6 = day6;
var dates = tslib_1.__importStar(require("./dates"));
exports.dates = dates;
var dynamic = tslib_1.__importStar(require("./dynamic"));
exports.dynamic = dynamic;
var generateSchema = tslib_1.__importStar(require("./generate"));
exports.generateSchema = generateSchema;
var generateUISchema = tslib_1.__importStar(require("./generateUI"));
exports.generateUISchema = generateUISchema;
var layout = tslib_1.__importStar(require("./layout"));
exports.layout = layout;
var person = tslib_1.__importStar(require("./person"));
exports.person = person;
var rule = tslib_1.__importStar(require("./rule"));
exports.rule = rule;
var ruleInheritance = tslib_1.__importStar(require("./ruleInheritance"));
exports.ruleInheritance = ruleInheritance;
var resolve = tslib_1.__importStar(require("./resolve"));
exports.resolve = resolve;
var config = tslib_1.__importStar(require("./config"));
exports.config = config;
var text = tslib_1.__importStar(require("./text"));
exports.text = text;
var numbers = tslib_1.__importStar(require("./numbers"));
exports.numbers = numbers;
var scope = tslib_1.__importStar(require("./scope"));
exports.scope = scope;
var listWithDetail = tslib_1.__importStar(require("./list-with-detail"));
exports.listWithDetail = listWithDetail;
var listWithDetailRegistered = tslib_1.__importStar(require("./list-with-detail-registered"));
exports.listWithDetailRegistered = listWithDetailRegistered;
var object = tslib_1.__importStar(require("./object"));
exports.object = object;
var i18n = tslib_1.__importStar(require("./i18n"));
exports.i18n = i18n;
var issue_1169 = tslib_1.__importStar(require("./1169"));
exports.issue_1169 = issue_1169;
var issue_1220 = tslib_1.__importStar(require("./1220"));
exports.issue_1220 = issue_1220;
var issue_1253 = tslib_1.__importStar(require("./1253"));
exports.issue_1253 = issue_1253;
var issue_1254 = tslib_1.__importStar(require("./1254"));
exports.issue_1254 = issue_1254;
var oneOfRecursive = tslib_1.__importStar(require("./oneOf-recursive"));
exports.oneOfRecursive = oneOfRecursive;
var huge = tslib_1.__importStar(require("./huge"));
exports.huge = huge;
var defaultExample = tslib_1.__importStar(require("./default"));
exports.defaultExample = defaultExample;
var onChange = tslib_1.__importStar(require("./onChange"));
exports.onChange = onChange;
var enumExample = tslib_1.__importStar(require("./enum"));
exports.enumExample = enumExample;
tslib_1.__exportStar(require("./register"), exports);
tslib_1.__exportStar(require("./example"), exports);
var ifThenElse = tslib_1.__importStar(require("./if_then_else"));
exports.ifThenElse = ifThenElse;
